src/codegen/dispatch/substrate.rs

```diff
@@ -0,0 +1,384 @@
+        let case_bb = self.cfg.new_basic_block(format!("msg_{msg_no}_dispatch"));
+        self.cfg.set_basic_block(case_bb);
+        self.abort_if_value_transfer(msg_no);
+
+        // Decode input data if necessary
+        let cfg = &self.all_cfg[msg_no];
+        let mut args = vec![];
+        if !cfg.params.is_empty() {
+            let buf_len = Expression::Variable(Codegen, Uint(32), self.input_len);
+            let arg_len = Expression::Subtract(
+                Codegen,
+                Uint(32),
+                false,
+                buf_len.into(),
+                self.selector_len.clone(),
+            );
+            args = abi_decode(
+                &Codegen,
+                &self.input_ptr,
+                &cfg.params.iter().map(|p| p.ty.clone()).collect::<Vec<_>>(),
+                self.ns,
+                &mut self.vartab,
+                &mut self.cfg,
+                Some(Expression::Trunc(Codegen, Uint(32), arg_len.into())),
+            );
+        }
+
+        let mut returns: Vec<usize> = Vec::with_capacity(cfg.returns.len());
+        let mut return_tys: Vec<Type> = Vec::with_capacity(cfg.returns.len());
+        let mut returns_expr: Vec<Expression> = Vec::with_capacity(cfg.returns.len());
+        for item in cfg.returns.iter() {
+            let new_var = self.vartab.temp_anonymous(&item.ty);
+            returns.push(new_var);
+            return_tys.push(item.ty.clone());
+            returns_expr.push(Expression::Variable(Codegen, item.ty.clone(), new_var));
+        }
+
+        self.add(Instr::Call {
+            res: returns,
+            call: InternalCallTy::Static { cfg_no: msg_no },
+            args,
+            return_tys,
+        });
+
+        if cfg.returns.is_empty() {
+            let data_len = Expression::NumberLiteral(Codegen, Uint(32), 0.into());
+            let data = Expression::AllocDynamicBytes(
+                Codegen,
+                Type::DynamicBytes,
+                data_len.clone().into(),
+                None,
+            );
+            self.add(Instr::ReturnData { data, data_len })
+        } else {
+            let (data, data_len) = abi_encode(
+                &Codegen,
+                returns_expr,
+                self.ns,
+                &mut self.vartab,
+                &mut self.cfg,
+                false,
+            );
+            self.add(Instr::ReturnData { data, data_len });
+        }
+
+        case_bb
+    }
+
+    /// Insert a trap into the cfg, if the message `msg_no` is not payable but received value anyways.
+    /// Constructors always receive endowment.
+    fn abort_if_value_transfer(&mut self, msg_no: usize) {
+        if !self.all_cfg[msg_no].nonpayable || self.all_cfg[msg_no].ty == FunctionTy::Constructor {
+            return;
+        }
+
+        let true_block = self.cfg.new_basic_block(format!("msg_{msg_no}_has_value"));
+        let false_block = self.cfg.new_basic_block(format!("msg_{msg_no}_no_value"));
```

Comment:
Why are you calling functions as messages?

Reply:
You are right about this. I used message because that's how callable [methods are called in ink!](https://use.ink/macros-attributes/message). However this is a solidity compiler so we are dealing with functions and not messages. I'm going to change that.